pkg/azure/instrument/instrument.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2023 SAP SE or an SAP affiliate company
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package instrument
+
+import (
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machinecodes/status"
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/metrics"
+)
+
+const prometheusProviderLabelValue = "azure"
+
+// RecordAzAPIMetric records a prometheus metric for Azure API calls.
+// * If there is an error then it will increment the APIFailedRequestCount counter vec metric.
+// * If the Azure API call is successful then it will record 2 metrics:
+//   - It will increment APIRequestCount counter vec metric. (Ideally this should have been named as APISuccessRequestCount)
+//   - It will compute the time taken for API call completion and record it.
+//
+// NOTE: If this function is called via `defer` then please keep in mind that parameters passed to defer are evaluated at the time of definition.
+// So if you have an error that is computed later in the function then ensure that you use named return parameters.
+func RecordAzAPIMetric(err error, azServiceName string, invocationTime time.Time) {
```

Comment:
For the possible values of `azServiceName` or `operation`, should we maintain a doc in this repo , or does the prometheus show the possible values, while writing the query?

Reply:
Unfortunately i have not been able to test the prometheus metrics part yet and would need to test this locally. Once i see these metrics then i can better comment on this point.

Reply:
We found that prometheus list the possible values , as suggestion, so we won't need a doc